### DISCUSSION TITLE
Can we expand x axis string values across column/tick boundary?

### DISCUSSION THREAD

**Discussion by kelvin-ngure** [patrykandpatrick/vico#850] [Questions] [answered]
### Question

Given the below screen, when I set `zoomState` to `Zoom.Content` so that my user doesn't have to scroll to see all the data, the values on the x axis shrink to 3 dots `...` likely because of all the squeezing. I also noticed they disappeared on my phone when `columnCollectionSpacing` was low (around 13-17 .dp)
![Screenshot_20240817_010929_Harrier](https://github.com/user-attachments/assets/fe9a73ab-6f1f-456e-aee7-c67e1d71422d)

Since my input values (shown below) are mostly empty strings, I was hoping that the ones that aren't empty would just occupy more space and be visible. How can I make this happen?

![image](https://github.com/user-attachments/assets/82d7d9b7-ac0a-404c-ba29-5937d1c5b5f6)

### Vico version(s)

2.0.0-alpha.27

### UI framework(s)

Jetpack Compose

**Comment by kelvin-ngure** [comment, marked as answer]
I was able to make the numbers visible even with all the empty strings removed like below by using the `itemPlacer` param in the bottomAxis
```    val hoursOfDay = listOf(
        "00:00", "01", "02", "03", "04", "05", "06:00", "07", "08", "09",
        "10", "11", "12", "13", "14", "15", "16", "17", "18:00", "19",
        "20", "21", "22", "23:00", "24:00"
    )
```

I ended up having an issue where the first and last values in the x axis were not visible. The instructions in [these release notes](https://github.com/patrykandpatrick/vico/releases/tag/v1.13.0-alpha.9) helped me get the first value visible but no matter what I put as the `scalableEndPadding`, I couldn't get the last value visible (I only added '24:00' for testing so that it fits the spacing of 6 but it would be great to know how I can get 0,6,12,18, and then 23)

The below screenshot is based on this code
```
bottomAxis = rememberBottomAxis(
                        label= rememberAxisLabelComponent(padding = Dimensions.of(0.dp)),
                        guideline = null,
                        valueFormatter = valueFormatter,
                        tickLength = 3.dp,
                        itemPlacer = remember { HorizontalAxis.ItemPlacer.default(spacing = 6, offset = 0, addExtremeLabelPadding = true)}
                    ),
                    horizontalLayout = HorizontalLayout.fullWidth(scalableStartPadding = 96.dp, scalableEndPadding = 200.dp)
                ),
                modelProducer = modelProducer,
                zoomState = rememberVicoZoomState(maxZoom = Zoom.Content)
```

![Screenshot_20240818_013817_Harrier](https://github.com/user-attachments/assets/f0aa7a00-d54b-4ea8-bfbf-8c0645968792)

**Reply by joseandroidengineer** [reply]
Thank you for this. This really helped me out!

**Reply by Gowsky** [reply]
Hello, @kelvin-ngure.

> I was able to make the numbers visible even with all the empty strings removed like below by using the `itemPlacer` param in the bottomAxis
> ```    val hoursOfDay = listOf(
>         "00:00", "01", "02", "03", "04", "05", "06:00", "07", "08", "09",
>         "10", "11", "12", "13", "14", "15", "16", "17", "18:00", "19",
>         "20", "21", "22", "23:00", "24:00"
>     )
> ```

Using `ItemPlacer` is the correct approach. You should never return empty strings from a `CartesianValueFormatter` used for an axis in an attempt to control the label spacing. An empty label is still a label, so it limits the space available to adjacent labels. `2.0.0-alpha.1` introduces a check that throws an exception when you try to use empty strings instead of `ItemPlacer`.

Note that `hoursOfDay` doesn't need to include strings like "01" and "02." You can leave just the ones that are displayed and make the rest empty. It doesn't matter, since they're never read.

That said, the solution with a list of labels is generally suboptimal here. What you should do is actually format the _x_ values. If the _x_ values are 0 through 23, then you can use something like this:

```kt
CartesianValueFormatter { x, _, _ ->
    "${x.toInt().toString().padStart(length = 2, padChar = '0')}:00"
}
```

If they're timestamps, you can use `DateTimeFormatter` or similar. A list of labels should be used only when you have arbitrary categories, not in numerical contexts.


> I ended up having an issue where the first and last values in the x axis were not visible. The instructions in [these release notes](https://github.com/patrykandpatrick/vico/releases/tag/v1.13.0-alpha.9) helped me get the first value visible but no matter what I put as the `scalableEndPadding`, I couldn't get the last value visible (I only added '24:00' for testing so that it fits the spacing of 6

It sounds like 24:00 is outside of the _x_ range of the chart. Simply adding such a label to the list will have no effect. You have to adjust the maximum _x_ as required via `AxisValueOverrider` (`CartesianLayerRangeProvider` in newer versions).

> but it would be great to know how I can get 0,6,12,18, and then 23)

You can do this by creating a custom `HorizontalAxis.ItemPlacer` implementation. You can delegate to `HorizontalAxis.ItemPlacer.default(...)` and just make these changes:

- Override `getLastLabelValue` so that it returns the _x_ value corresponding to 23:00.
- Override `getLabelValues` so that it adds the _x_ value corresponding to 23:00 to the list returned by `super.getLabelValues`.
